sw/device/silicon_creator/mask_rom/bootstrap.c

```diff
@@ -0,0 +1,321 @@
+  });
+  rom_error_t error = flash_ctrl_data_write(addr, word_count, data);
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolFalse,
+  });
+  return error;
+}
+
+/**
+ * Bootstrap state 1: Wait for an erase command and erase the data
+ * partition.
+ *
+ * This function erases both data banks of the flash regardless of the type of
+ * the erase command (CHIP_ERASE or SECTOR_ERASE).
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_erase(bootstrap_state_t *state) {
+  HARDENED_CHECK_EQ(*state, kBootstrapStateErase);
+
+  spi_device_cmd_t cmd;
+  spi_device_cmd_get(&cmd);
+  rom_error_t error = kErrorUnknown;
+  switch (cmd.opcode) {
+    case kSpiDeviceOpcodeChipErase:
+    case kSpiDeviceOpcodeSectorErase:
+      error = bootstrap_chip_erase();
+      HARDENED_RETURN_IF_ERROR(error);
+      *state = kBootstrapStateEraseVerify;
+      break;
+    default:
+      // Ignore any other command, e.g. PAGE_PROGRAM, RESET.
+      error = kErrorOk;
+  }
+
+  // Note: We clear WIP and WEN bits in `bootstrap_handle_erase_verify()` after
+  // checking that both data banks have been erased.
+  return error;
+}
+
+/**
+ * Bootstrap state 2: Verify that all data banks have been erased.
+ *
+ * This function also clears the WIP and WEN bits of the flash status register.
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_erase_verify(bootstrap_state_t *state) {
+  HARDENED_CHECK_EQ(*state, kBootstrapStateEraseVerify);
+
+  rom_error_t err_0 = flash_ctrl_data_erase_verify(0, kFlashCtrlEraseTypeBank);
+  rom_error_t err_1 = flash_ctrl_data_erase_verify(
+      FLASH_CTRL_PARAM_BYTES_PER_BANK, kFlashCtrlEraseTypeBank);
+  HARDENED_RETURN_IF_ERROR(err_0);
+  HARDENED_RETURN_IF_ERROR(err_1);
+
+  *state = kBootstrapStateProgram;
+  spi_device_flash_status_clear();
+  return err_0;
+}
+
+/**
+ * Bootstrap state 3: (Erase/)Program loop.
+ *
+ * @param state Bootstrap state.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_handle_program(bootstrap_state_t *state) {
+  static_assert(alignof(spi_device_cmd_t) >= sizeof(uint32_t) &&
+                    offsetof(spi_device_cmd_t, payload) >= sizeof(uint32_t),
+                "Payload must be word aligned.");
+
+  HARDENED_CHECK_EQ(*state, kBootstrapStateProgram);
+
+  spi_device_cmd_t cmd;
+  spi_device_cmd_get(&cmd);
+  rom_error_t error = kErrorUnknown;
+  switch (cmd.opcode) {
+    case kSpiDeviceOpcodeChipErase:
+      error = bootstrap_chip_erase();
+      break;
+    case kSpiDeviceOpcodeSectorErase:
+      error = bootstrap_sector_erase(cmd.address);
+      break;
+    case kSpiDeviceOpcodePageProgram:
+      error = bootstrap_page_program(cmd.address, cmd.payload_byte_count,
+                                     cmd.payload);
+      break;
```

Comment:
We probably ought to check that WEN has been set before proceeding to do the state-changing operations.

Reply:
Should we ignore the command if WEN is not set or should this be an error?

Reply:
An ordinary flash part would just ignore the command and carry on, but I think there might be good reasons to treat it as an error. This should be a fairly controlled operation, so perhaps providing warning of any issues and shutting down a bad process is the way to go. :thinking: 

I'm not really sure, though.

Reply:
Done, we now ignore any erase or program command if WEL is not set. Please note that this does not apply to reset commands.